GVFS/FastFetch/CheckoutPrefetcher.cs

```diff
@@ -72,87 +72,104 @@ public override void Prefetch(string branchOrCommit, bool isBranch)
                     {
-                        Index newIndex = new Index(
-                            this.Enlistment.EnlistmentRoot,
-                            this.Tracer,
-                            Path.Combine(this.Enlistment.DotGitRoot, GVFSConstants.DotGit.IndexName),
-                            readOnly: false);
-
-                        // Update from disk only if the caller says it is ok via command line
-                        // or if we updated the whole tree and know that all files are up to date
-                        bool allowIndexMetadataUpdateFromWorkingTree = this.allowIndexMetadataUpdateFromWorkingTree || checkout.UpdatedWholeTree;
-                        newIndex.UpdateFileSizesAndTimes(checkout.AddedOrEditedLocalFiles, allowIndexMetadataUpdateFromWorkingTree, shouldSignIndex, sourceIndex);
+                        Index sourceIndex = this.GetSourceIndex();
+                        GitIndexGenerator indexGen = new GitIndexGenerator(this.Tracer, this.Enlistment, shouldSignIndex);
+                        indexGen.CreateFromHeadTree(indexVersion: 2);
+                        this.HasFailures |= indexGen.HasFailures;
+
+                        if (!indexGen.HasFailures)
+                        {
+                            Index newIndex = new Index(
+                                this.Enlistment.EnlistmentRoot,
+                                this.Tracer,
+                                Path.Combine(this.Enlistment.DotGitRoot, GVFSConstants.DotGit.IndexName),
+                                readOnly: false);
+
+                            // Update from disk only if the caller says it is ok via command line
+                            // or if we updated the whole tree and know that all files are up to date
+                            bool allowIndexMetadataUpdateFromWorkingTree = this.allowIndexMetadataUpdateFromWorkingTree || checkout.UpdatedWholeTree;
+                            newIndex.UpdateFileSizesAndTimes(checkout.AddedOrEditedLocalFiles, allowIndexMetadataUpdateFromWorkingTree, shouldSignIndex, sourceIndex);
+                        }
                     }
                 }
+
+                indexLock.Release();
+                // Note that this releases the lock even if we had failures, and thus even if the index/working-tree
+                // might need to be repaired.  It is up to our caller to monitor the exit code and repair/discard the
+                // enlistment as-appropriate.  If we have an unhandled exception or are killed, the file will be left
+                // around as well.  The thinking here is that if we are killed, it's like our parent is too and thus
+                // will not be able to take those recovery steps, and leaving the file around ends up being a flag to
+                // the next command that's run after the machine recovers to sort it out.

```

Comment:
What I was going for was to leave the index.lock file around in case fastfetch got interrupted, so we'd know to wipe the machine.  I wasn't able to convince the build lab that this was the way to go, at least not today.  This iteration attempts to smash all the file-move operations as close together as possible to reduce the chance it'll get left in a half-baked state.